### PR TITLE
Proposed 0.0.3: coordinate system & camera control

### PR DESCRIPTION
**Pull request by victoriacity**
The following features are added in this fork based on the `taichi_three:dev` branch:

Implemented full camera control: use scene.camera.from_mouse(gui) after a gui.get_event() call (without arguments) to control the camera with the mouse.
* Drag with the left mouse button pressed to orbit the camera.
* Drag with the right mouse button or use the wheel to zoom in and out.
* Drag with the middle mouse buttom to pan the camera.

Implemented Backface culling for perspective cameras

Used the area method to compute barycentric coordinates

Clipped the triangle bounding box to prevent drawing outside the camera

Updated the `monkey.obj` asset to the correct orientation using +Z forward, +Y up

Fixed SceneRT compatibility with the refactor API for the `dev` branch

### CODE REVIEW

File: taichi_three/version.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
-version = (0, 0, 2)
+version = (0, 0, 4)
```
Comment (archibate) [inline taichi_three/version.py:1]:
```suggestion
version = (0, 0, 3)
```

File: taichi_three/transform.py [hunk 2]
```diff
@@ -133,23 +133,27 @@ class Camera(AutoInit):
...
         target = self.target_py if target is None else target
         up = self.up_py if up is None else up
+        
         # fwd = target - pos
         fwd = [target[i] - pos[i] for i in range(3)]
         # fwd = fwd.normalized()
         fwd_len = math.sqrt(sum(x**2 for x in fwd))
         fwd = [x / fwd_len for x in fwd]
-        # right = fwd.cross(up)
+        # right = fwd.cross(up)
```
Comment (archibate) [inline taichi_three/transform.py:151]:
```suggestion
        # right = fwd.cross(up)
```

File: README.md [hunk 3]
```diff
@@ -10,6 +10,20 @@ Taichi THREE is an extension library of the [Taichi Programming Language](https:
 ![Example](https://github.com/taichi-dev/taichi_three/raw/master/assets/monkey.png)
 
 
+New in 0.0.4
```
Comment (archibate) [inline README.md:13]:
```suggestion
New in 0.0.3
```

**Comment by archibate** [comment]
It seems that if I scale first, then **the orbiting center** is changed?

File: taichi_three/transform.py [hunk 4]
```diff
@@ -168,16 +190,73 @@ def clear_buffer(self):
...
-        if isinstance(mpos, ti.GUI):
-            mpos = mpos.get_cursor_pos()
-
-        a, t = mpos
-        if a != 0 or t != 0:
-            a, t = a * math.tau - math.pi, t * math.pi - math.pi / 2
-        d = dis * math.cos(t)
-        self.set(pos=[d * math.sin(a), dis * math.sin(t), -d * math.cos(a)])
-        self._init()
+    def from_mouse(self, gui):
```
Comment (victoriacity) [inline taichi_three/transform.py:193]:
```suggestion
    def from_mouse(self, gui):
        is_alter_move = gui.is_pressed(ti.GUI.CTRL)
```

File: taichi_three/transform.py [hunk 5]
```diff
@@ -168,16 +190,73 @@ def clear_buffer(self):
...
+        if gui.is_pressed(ti.GUI.LMB):
+            mpos = gui.get_cursor_pos()
+            if self.mpos != (0, 0):
+                self.orbit((mpos[0] - self.mpos[0], mpos[1] - self.mpos[1]),
+                    pov=gui.is_pressed(ti.GUI.CTRL))
```
Comment (victoriacity) [inline taichi_three/transform.py:198]:
```suggestion
                    pov=is_alter_move)
```

File: taichi_three/transform.py [hunk 6]
```diff
@@ -168,16 +190,73 @@ def clear_buffer(self):
...
+            self.mpos = mpos
+        elif gui.is_pressed(ti.GUI.RMB):
+            mpos = gui.get_cursor_pos()
+            if self.mpos != (0, 0):
+                self.zoom_by_mouse(mpos, (mpos[0] - self.mpos[0], mpos[1] - self.mpos[1]))
```
Comment (victoriacity) [inline taichi_three/transform.py:203]:
```suggestion
                self.zoom_by_mouse(mpos, (mpos[0] - self.mpos[0], mpos[1] - self.mpos[1]),
                        dolly=is_alter_move)
```

File: taichi_three/transform.py [hunk 7]
```diff
@@ -168,16 +190,73 @@ def clear_buffer(self):
...
+            self.mpos = mpos
+        elif gui.is_pressed(ti.GUI.MMB):
+            mpos = gui.get_cursor_pos()
+            if self.mpos != (0, 0):
+                self.pan((mpos[0] - self.mpos[0], mpos[1] - self.mpos[1]))
+            self.mpos = mpos
+        else:
+            if gui.event and gui.event.key == ti.GUI.WHEEL:
+                # one mouse wheel unit is (0, 120)
+                self.zoom(-gui.event.delta[1] / 1200)
```
Comment (victoriacity) [inline taichi_three/transform.py:213]:
```suggestion
                self.zoom(-gui.event.delta[1] / 1200,
                    dolly=is_alter_move)
```

File: taichi_three/transform.py [hunk 8]
```diff
@@ -168,16 +190,73 @@ def clear_buffer(self):
...
+        if ds != 0 or dt != 0:
+            z = math.sqrt(ds ** 2 + dt ** 2) * sensitivity
+            if (pos[0] - 0.5) * ds + (pos[1] - 0.5) * dt > 0:
+                z *= -1
+            self.zoom(z)
+    
+    def zoom(self, z):
+        newpos = [(1 + z) * self.pos_py[i] - z * self.target_py[i] for i in range(3)]
+        newtarget = [z * self.pos_py[i] + (1 - z) * self.target_py[i] for i in range(3)]
+        self.set(pos=newpos, target=newtarget)
```
Comment (victoriacity) [inline taichi_three/transform.py:246]:
```suggestion
    def zoom(self, z, dolly=False):
        newpos = [(1 + z) * self.pos_py[i] - z * self.target_py[i] for i in range(3)]
        if dolly:
            newtarget = [z * self.pos_py[i] + (1 - z) * self.target_py[i] for i in range(3)]
            self.set(pos=newpos, target=newtarget)
        else:
            self.set(pos=newpos)
```

**Comment by archibate** [comment]
The target should not be changed without pressing CTRL, IMO. This allows users to orbit the target in a closer details when zoomed in.
Btw, would it be nice add some hints in `from_mouse`'s first invocation, showing something like `[Tai3D] Hint: use LMB to orbit, RMB to ...`? So that users know how to control our cool camera system.

**Pull request merged**